spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/BaseColumnBatchLoader.java

```diff
@@ -0,0 +1,199 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.data.vectorized;
+
+import java.util.Iterator;
+import org.apache.iceberg.data.DeleteFilter;
+import org.apache.iceberg.deletes.PositionDeleteIndex;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.Pair;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.vectorized.ColumnVector;
+import org.apache.spark.sql.vectorized.ColumnarBatch;
+
+@SuppressWarnings("checkstyle:VisibilityModifier")
+public abstract class BaseColumnBatchLoader {
+  protected final int numRowsToRead;
+  // the rowId mapping to skip deleted rows for all column vectors inside a batch, it is null when
+  // there is no deletes
+  protected int[] rowIdMapping;
+  // the array to indicate if a row is deleted or not, it is null when there is no "_deleted"
+  // metadata column
+  protected boolean[] isDeleted;
+  private final boolean hasIsDeletedColumn;
+  private final DeleteFilter<InternalRow> deletes;
+  private final long rowStartPosInBatch;
+
+  protected BaseColumnBatchLoader(
+      int numRowsToRead,
+      boolean hasIsDeletedColumn,
+      DeleteFilter<InternalRow> deletes,
+      long rowStartPosInBatch) {
+    Preconditions.checkArgument(
+        numRowsToRead > 0, "Invalid number of rows to read: %s", numRowsToRead);
+    this.numRowsToRead = numRowsToRead;
+    this.hasIsDeletedColumn = hasIsDeletedColumn;
+    this.deletes = deletes;
+    this.rowStartPosInBatch = rowStartPosInBatch;
+    if (hasIsDeletedColumn) {
+      isDeleted = new boolean[numRowsToRead];
+    }
+  }
+
+  public ColumnarBatch loadDataToColumnBatch() {
+    int numRowsUndeleted = initRowIdMapping();
+
+    ColumnVector[] arrowColumnVectors = readDataToColumnVectors();
+
+    ColumnarBatch newColumnarBatch = new ColumnarBatch(arrowColumnVectors);
+    newColumnarBatch.setNumRows(numRowsUndeleted);
+
+    if (hasEqDeletes()) {
+      applyEqDelete(newColumnarBatch);
+    }
+
+    if (hasIsDeletedColumn && rowIdMapping != null) {
+      // reset the row id mapping array, so that it doesn't filter out the deleted rows
+      for (int i = 0; i < numRowsToRead; i++) {
+        rowIdMapping[i] = i;
+      }
+      newColumnarBatch.setNumRows(numRowsToRead);
+    }
+
+    if (hasIsDeletedColumn) {
+      readDeletedColumnIfNecessary(arrowColumnVectors);
+    }
+
+    return newColumnarBatch;
+  }
+
+  protected abstract ColumnVector[] readDataToColumnVectors();
+
+  protected abstract void readDeletedColumnIfNecessary(ColumnVector[] arrowColumnVectors);
```

Comment:
I think we need some more documentation on these abstract methods, especially since our base implementation doesn't use them

Reply:
Thinking more on this, should the base implementation be empty if it's not required?

Reply:
I have added some documentation and also refactored the code a bit, so we don't need this comet specific code in base class any more.